File: rxlocationmanager/src/main/java/ru/solodovnikov/rxlocationmanager/RxLocationManager.java
package ru.solodovnikov.rxlocationmanager;

import android.annotation.TargetApi;
import android.content.Context;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Build;
import android.os.Bundle;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.concurrent.Callable;
import java.util.concurrent.TimeoutException;

import ru.solodovnikov.rxlocationmanager.error.ElderLocationException;
import ru.solodovnikov.rxlocationmanager.error.ProviderDisabledException;
import rx.Observable;
import rx.Scheduler;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;


public class RxLocationManager {

    private final LocationManager locationManager;
    private final Scheduler scheduler;

    public RxLocationManager(Context context) {
        this.locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        this.scheduler = AndroidSchedulers.mainThread();
    }

    RxLocationManager(LocationManager locationManager, Scheduler scheduler) {
        this.locationManager = locationManager;
        this.scheduler = scheduler;
    }

    /**
     * Get last location from specific provider
     * Observable will emit ElderLocationException if {@code howOldCanBe} is not null and location time is not valid.
     *
     * @param provider    provider name
     * @param howOldCanBe optional. How old a location can be
     * @return observable that emit last known location. May emit null.
     * @see ElderLocationException
     */
    public Observable<Location> getLastLocation(final @NonNull String provider, final @Nullable LocationTime howOldCanBe) {
        return Observable.fromCallable(new Callable<Location>() {
            @Override
            public Location call() throws Exception {
                try {
                    return locationManager.getLastKnownLocation(provider);
                } catch (SecurityException ex) {
                    throw ex;
                }
            }
        }).compose(new Observable.Transformer<Location, Location>() {
            @Override
            public Observable<Location> call(Observable<Location> locationObservable) {

                final Observable<Location> filterObservable = locationObservable.map(new Func1<Location, Location>() {
                    @Override
                    public Location call(Location location) {
                        if (location != null && !isLocationNotOld(location, howOldCanBe)) {
                            throw new ElderLocationException(location);
                        }

                        return location;
                    }
                });

                return howOldCanBe != null ? filterObservable : locationObservable;
            }
        }).compose(applySchedulers());
    }


    /**
     * Get last location from specific provider
     *
     * @param provider provider name
     * @return observable that emit last known location. May emit null.
     */
    public Observable<Location> getLastLocation(@NonNull String provider) {
        return getLastLocation(provider, null);
    }

    /**
     * Try to get current location by specific provider.
     * Observable will emit TimeoutException in case of timeOut if timeOut object is not null.
     * Observable will emit ProviderDisabledException if provider is disabled
     *
     * @param provider provider name
     * @param timeOut  optional request timeout
     * @return observable that emit current location
     * @see TimeoutException
     * @see ProviderDisabledException
     */
    public Observable<Location> requestLocation(final @NonNull String provider, final @Nullable LocationTime timeOut) {
        return requestLocation(provider, timeOut, true);
    }

    /**
     * Try to get current location by specific provider
     * Observable will emit ProviderDisabledException if provider is disabled
     *
     * @param provider provider name
     * @return observable that emit current location
     * @see ProviderDisabledException
     */
    public Observable<Location> requestLocation(@NonNull String provider) {
        return requestLocation(provider, null);
    }

    Observable<Location> requestLocation(final @NonNull String provider, final @Nullable LocationTime timeOut, boolean throwExceptionIfDisabled) {

        final RxLocationListener locationListener = new RxLocationListener(locationManager, provider, throwExceptionIfDisabled);

        return Observable.create(locationListener)
                .compose(new Observable.Transformer<Location, Location>() {
                    @Override
                    public Observable<Location> call(Observable<Location> locationObservable) {
                        return timeOut != null ? locationObservable.timeout(timeOut.getValue(), timeOut.getTimeUnit()) : locationObservable;
                    }
                })
                .compose(applySchedulers());
    }

    private Observable.Transformer<Location, Location> applySchedulers() {
        return new Observable.Transformer<Location, Location>() {
            @Override
            public Observable<Location> call(Observable<Location> locationObservable) {
                return locationObservable.subscribeOn(scheduler).observeOn(scheduler);
            }
        };
    }

    private boolean isLocationNotOld(Location location, LocationTime howOldCanBe) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            return isLocationNotOld17(location, howOldCanBe);
        } else {
            return isLocationNotOldDefault(location, howOldCanBe);
        }
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    private boolean isLocationNotOld17(Location location, LocationTime howOldCanBe) {
        return SystemClock.elapsedRealtimeNanos() - location
                .getElapsedRealtimeNanos() < howOldCanBe.getTimeUnit().toNanos(howOldCanBe.getValue());
    }

    private boolean isLocationNotOldDefault(Location location, LocationTime howOldCanBe) {
        return System.currentTimeMillis() - location.getTime() < howOldCanBe.getTimeUnit().toMillis(howOldCanBe.getValue());
    }

    private static class RxLocationListener implements Observable.OnSubscribe<Location> {

        private final LocationManager locationManager;
        private final String provider;
        private final boolean throwExceptionIfDisabled;

        private LocationListener locationListener;

        public RxLocationListener(LocationManager locationManager, String provider, boolean throwExceptionIfDisabled) {
            this.locationManager = locationManager;
            this.provider = provider;
            this.throwExceptionIfDisabled = throwExceptionIfDisabled;
        }

        @Override
        public void call(final Subscriber<? super Location> subscriber) {
            if (locationManager.isProviderEnabled(provider)) {
                locationListener = new LocationListener() {
                    @Override
                    public void onLocationChanged(Location location) {
                        if (subscriber != null && !subscriber.isUnsubscribed()) {
                            subscriber.onNext(location);
                            subscriber.onCompleted();
                        }
                    }

                    @Override
                    public void onStatusChanged(String provider, int status, Bundle extras) {

                    }

                    @Override
                    public void onProviderEnabled(String provider) {

                    }

                    @Override
                    public void onProviderDisabled(String provider) {
                        if (RxLocationListener.this.provider.equals(provider)) {
                            subscriber.onError(new ProviderDisabledException(provider));
                        }
                    }
                };
                try {
                    locationManager.requestSingleUpdate(provider, locationListener, null);
                } catch (SecurityException ex) {
                    throw ex;
                }
                subscriber.add(new Subscription() {
                    @Override
                    public void unsubscribe() {
                        if (!subscriber.isUnsubscribed()) {
                            subscriber.unsubscribe();
                        }

                        removeUpdates();
                    }

                    @Override
                    public boolean isUnsubscribed() {
                        return subscriber.isUnsubscribed();
                    }
                });
            } else {

                if (throwExceptionIfDisabled) {
                    subscriber.onError(new ProviderDisabledException(provider));
                } else {
                    subscriber.onCompleted();
                }
            }
        }

        void removeUpdates() {
            if (locationListener != null) {
                try {
                    locationManager.removeUpdates(locationListener);
                } catch (SecurityException ex) {
                    throw ex;
                }
            }
        }
    }
}

File: rxlocationmanager/src/main/kotlin/ru/solodovnikov/rxlocationmanager/RxLocationManager.kt
package ru.solodovnikov.rxlocationmanager

import android.annotation.SuppressLint
import android.app.Instrumentation
import android.content.Context
import android.content.Intent
import android.location.*
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import rx.Emitter
import rx.Observable
import rx.Scheduler
import rx.Single
import rx.android.schedulers.AndroidSchedulers
import rx.subjects.PublishSubject
import java.util.concurrent.TimeoutException

/**
 * Implementation of [BaseRxLocationManager] based on RxJava1
 */
@SuppressLint("MissingPermission")
class RxLocationManager internal constructor(context: Context,
                                             private val scheduler: Scheduler
) : BaseRxLocationManager(context) {

    constructor(context: Context) : this(context, AndroidSchedulers.mainThread())

    private val permissionSubject by lazy { PublishSubject.create<Pair<Array<out String>, IntArray>>() }
    private val resultSubject by lazy { PublishSubject.create<Instrumentation.ActivityResult>() }

    /**
     * Get last location from specific provider
     * Observable will emit [ElderLocationException] if [howOldCanBe] is not null and location time is not valid.
     *
     * @param provider provider name
     * @param howOldCanBe how old a location can be
     * @param behaviors extra behaviors
     * @return observable that emit last known location
     * @see ElderLocationException
     */
    @JvmOverloads
    fun getLastLocation(provider: String,
                        howOldCanBe: LocationTime? = null,
                        vararg behaviors: SingleBehavior): Single<Location> =
            Single.fromCallable { locationManager.getLastKnownLocation(provider) }
                    .compose {
                        if (howOldCanBe != null) {
                            it.doOnSuccess {
                                if (it != null && !it.isNotOld(howOldCanBe)) {
                                    throw ElderLocationException(it)
                                }
                            }
                        } else {
                            it
                        }
                    }.applyBehaviors(behaviors, BehaviorParams(provider))
                    .compose(this::applySchedulers)

    /**
     * Emit any [LocationListener] events. Observable will completed as soon as it emits any location.
     * Observable will emit [TimeoutException] if [timeOut] is not null and timeOut occurs.
     *
     * @param provider provider name
     * @param timeOut  request timeout
     * @param behaviors extra behaviors
     * @return [Observable] that emit location events
     * @see TimeoutException
     * @see BaseRxLocationManager.LocationEvent
     */
    @JvmOverloads
    fun requestLocationRaw(provider: String,
                           timeOut: LocationTime? = null,
                           vararg behaviors: ObservableBehavior): Observable<LocationEvent> =
            Observable.create<LocationEvent>({ emitter ->
                SingleLocationListener(emitter).also {
                    emitter.setCancellation { locationManager.removeUpdates(it) }

                    locationManager.requestSingleUpdate(provider, it, null)
                }
            }, Emitter.BackpressureMode.NONE)
                    .compose { if (timeOut != null) it.timeout(timeOut.time, timeOut.timeUnit) else it }
                    .applyBehaviors(behaviors, BehaviorParams(provider))
                    .compose(this::applySchedulers)

    /**
     * Try to get and emit current location.
     * Single will emit [TimeoutException] if [timeOut] is not null and timeOut occurs.
     *
     * @param provider provider name
     * @param timeOut  request timeout
     * @param behaviors extra behaviors
     * @return [Single] that emit location events
     * @see TimeoutException
     */
    @JvmOverloads
    fun requestLocation(provider: String,
                        timeOut: LocationTime? = null,
                        vararg behaviors: SingleBehavior): Single<Location> =
            requestLocationRaw(provider, timeOut)
                    .first { it is LocationEvent.LocationChangedEvent }
                    .map { (it as LocationEvent.LocationChangedEvent).location }
                    .toSingle()
                    .applyBehaviors(behaviors, BehaviorParams(provider))

    /**
     * Register for any location listener updates
     *
     * @param provider the name of the provider with which to register
     * @param minTime minimum time interval between location updates, in milliseconds
     * @param minDistance minimum distance between location updates, in meters
     * @param behaviors extra behaviors
     * @return [Observable] that emit any [LocationListener] events
     */
    @JvmOverloads
    fun requestLocationUpdatesRaw(provider: String,
                                  minTime: Long,
                                  minDistance: Float = 0F,
                                  vararg behaviors: ObservableBehavior): Observable<LocationEvent> =
            Observable.create<LocationEvent>({ emitter ->
                DefaultLocationListener(emitter).also {
                    emitter.setCancellation { locationManager.removeUpdates(it) }

                    locationManager.requestLocationUpdates(provider, minTime, minDistance, it)
                }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams(provider))
                    .compose(this::applySchedulers)

    /**
     * Register for location updates
     *
     * @param provider the name of the provider with which to register
     * @param minTime minimum time interval between location updates, in milliseconds
     * @param minDistance minimum distance between location updates, in meters
     * @param behaviors extra behaviors
     * @return [Observable] that emit only location update events
     */
    @JvmOverloads
    fun requestLocationUpdates(provider: String,
                               minTime: Long = 0L,
                               minDistance: Float = 0.0F,
                               vararg behaviors: ObservableBehavior): Observable<Location> =
            requestLocationUpdatesRaw(provider, minTime, minDistance, *behaviors)
                    .filter { it is LocationEvent.LocationChangedEvent }
                    .map { (it as LocationEvent.LocationChangedEvent).location }

    /**
     * Adds a GPS status listener. Use [addGnssStatusListener] on API version >= N.
     *
     * @see LocationManager.addGpsStatusListener
     */
    @Suppress("DEPRECATION")
    fun addGpsStatusListener(vararg behaviors: ObservableBehavior): Observable<Int> =
            Observable.create<Int>({ emitter ->
                GpsStatus.Listener { event -> emitter.onNext(event) }.also {
                    emitter.setCancellation { locationManager.removeGpsStatusListener(it) }
                    if (!locationManager.addGpsStatusListener(it)) {
                        emitter.onError(ListenerNotRegisteredException())
                    }
                }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams())
                    .compose(this::applySchedulers)


    /**
     * Registers a GNSS status callback.
     *
     * @see LocationManager.registerGnssStatusCallback
     */
    @RequiresApi(Build.VERSION_CODES.N)
    fun addGnssStatusListener(vararg behaviors: ObservableBehavior): Observable<GnssStatusResponse> =
            Observable.create<GnssStatusResponse>({ emitter ->
                object : GnssStatus.Callback() {
                    override fun onSatelliteStatusChanged(status: GnssStatus) {
                        super.onSatelliteStatusChanged(status)
                        emitter.onNext(GnssStatusResponse.GnssSatelliteStatusChanged(status))
                    }

                    override fun onStopped() {
                        super.onStopped()
                        emitter.onNext(GnssStatusResponse.GnssStopped())
                    }

                    override fun onStarted() {
                        super.onStarted()
                        emitter.onNext(GnssStatusResponse.GnssStarted())
                    }

                    override fun onFirstFix(ttffMillis: Int) {
                        super.onFirstFix(ttffMillis)
                        emitter.onNext(GnssStatusResponse.GnssFirstFix(ttffMillis))
                    }
                }.also {
                    emitter.setCancellation { locationManager.unregisterGnssStatusCallback(it) }

                    if (!locationManager.registerGnssStatusCallback(it)) {
                        emitter.onError(ListenerNotRegisteredException())
                    }
                }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams())
                    .compose(this::applySchedulers)

    /**
     * Adds an NMEA listener.
     *
     * @see LocationManager.addNmeaListener
     */
    @RequiresApi(Build.VERSION_CODES.N)
    fun addNmeaListenerN(vararg behaviors: ObservableBehavior): Observable<NmeaMessage> =
            Observable.create<NmeaMessage>({ e ->
                OnNmeaMessageListener { message, timestamp -> e.onNext(NmeaMessage(message, timestamp)) }
                        .also {
                            e.setCancellation { locationManager.removeNmeaListener(it) }

                            if (!locationManager.addNmeaListener(it)) {
                                e.onError(ListenerNotRegisteredException())
                            }
                        }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams())
                    .compose(this::applySchedulers)

    /**
     * Adds an NMEA listener.
     *
     * @see LocationManager.addNmeaListener
     */
    @Suppress("DEPRECATION")
    fun addNmeaListener(vararg behaviors: ObservableBehavior): Observable<NmeaMessage> =
            Observable.create<NmeaMessage>({ e ->
                GpsStatus.NmeaListener { timestamp, nmea -> e.onNext(NmeaMessage(nmea, timestamp)) }
                        .also {
                            e.setCancellation { locationManager.removeNmeaListener(it) }

                            if (!locationManager.addNmeaListener(it)) {
                                e.onError(ListenerNotRegisteredException())
                            }

                        }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams())
                    .compose(this::applySchedulers)

    /**
     * Registers a GPS Measurement callback.
     *
     * @see LocationManager.registerGnssMeasurementsCallback
     */
    @RequiresApi(Build.VERSION_CODES.N)
    fun registerGnssMeasurementsCallback(vararg behaviors: ObservableBehavior): Observable<GnssMeasurementsResponse> =
            Observable.create<GnssMeasurementsResponse>({ emitter ->
                object : GnssMeasurementsEvent.Callback() {
                    override fun onGnssMeasurementsReceived(eventArgs: GnssMeasurementsEvent) {
                        super.onGnssMeasurementsReceived(eventArgs)
                        emitter.onNext(GnssMeasurementsResponse.GnssMeasurementsReceived(eventArgs))
                    }

                    override fun onStatusChanged(status: Int) {
                        super.onStatusChanged(status)
                        emitter.onNext(GnssMeasurementsResponse.StatusChanged(status))
                    }
                }.also {
                    emitter.setCancellation { locationManager.unregisterGnssMeasurementsCallback(it) }

                    if (!locationManager.registerGnssMeasurementsCallback(it)) {
                        emitter.onError(ListenerNotRegisteredException())
                    }
                }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams())
                    .compose(this::applySchedulers)

    /**
     * Registers a GNSS Navigation Message callback.
     *
     * @see LocationManager.registerGnssNavigationMessageCallback
     */
    @RequiresApi(Build.VERSION_CODES.N)
    fun registerGnssNavigationMessageCallback(vararg behaviors: ObservableBehavior): Observable<GnssNavigationResponse> =
            Observable.create<GnssNavigationResponse>({ emitter ->
                object : GnssNavigationMessage.Callback() {
                    override fun onGnssNavigationMessageReceived(event: GnssNavigationMessage) {
                        super.onGnssNavigationMessageReceived(event)
                        emitter.onNext(GnssNavigationResponse.GnssNavigationMessageReceived(event))
                    }

                    override fun onStatusChanged(status: Int) {
                        super.onStatusChanged(status)
                        emitter.onNext(GnssNavigationResponse.StatusChanged(status))
                    }
                }.also {
                    emitter.setCancellation { locationManager.unregisterGnssNavigationMessageCallback(it) }

                    if (!locationManager.registerGnssNavigationMessageCallback(it)) {
                        emitter.onError(ListenerNotRegisteredException())
                    }
                }
            }, Emitter.BackpressureMode.NONE)
                    .applyBehaviors(behaviors, BehaviorParams())
                    .compose(this::applySchedulers)

    /**
     * Retrieves information about the current status of the GPS engine.
     *
     * @see LocationManager.getGpsStatus
     */
    @JvmOverloads
    @Suppress("DEPRECATION")
    fun getGpsStatus(status: GpsStatus? = null) =
            Single.fromCallable { locationManager.getGpsStatus(status) }
                    .compose(this::applySchedulers)

    /**
     * Returns a list of the names of all known location providers.
     *
     * @see LocationManager.getAllProviders
     */
    fun getAllProviders() = Single.fromCallable { locationManager.allProviders }
            .compose(this::applySchedulers)

    /**
     * Returns a list of the names of location providers.
     *
     * @see LocationManager.getProviders
     */
    @JvmOverloads
    fun getProviders(criteria: Criteria? = null, enabledOnly: Boolean) =
            Single.fromCallable { locationManager.getProviders(criteria, enabledOnly) }
                    .compose(this::applySchedulers)

    /**
     * Returns the name of the provider that best meets the given criteria.
     *
     * @see LocationManager.getBestProvider
     */
    @JvmOverloads
    fun getBestProvider(criteria: Criteria, enabledOnly: Boolean = true) =
            Single.fromCallable { locationManager.getBestProvider(criteria, enabledOnly) }
                    .compose(this::applySchedulers)

    /**
     * Returns the information associated with the location provider of the given name, or null if no provider exists by that name.
     *
     * @param behaviors extra behaviors
     * @see LocationManager.getProvider
     */
    fun getProvider(name: String, vararg behaviors: SingleBehavior) =
            Single.fromCallable { locationManager.getProvider(name) }
                    .applyBehaviors(behaviors, BehaviorParams(name))

    /**
     * Returns the current enabled/disabled status of the given provider.
     *
     * @see LocationManager.isProviderEnabled
     */
    fun isProviderEnabled(provider: String, vararg behaviors: SingleBehavior): Single<Boolean> =
            Single.fromCallable { locationManager.isProviderEnabled(provider) }
                    .applyBehaviors(behaviors, BehaviorParams(provider))

    override fun onRequestPermissionsResult(permissions: Array<out String>, grantResults: IntArray) {
        permissionSubject.onNext(Pair(permissions, grantResults))
    }

    override fun onActivityResult(resultCode: Int, data: Intent?) {
        resultSubject.onNext(Instrumentation.ActivityResult(resultCode, data))
    }

    internal fun subscribeToPermissionUpdate(onUpdate: (Pair<Array<out String>, IntArray>) -> Unit)
            = permissionSubject.subscribe(onUpdate, {}, {})

    internal fun subscribeToActivityResultUpdate(onUpdate: (Instrumentation.ActivityResult) -> Unit)
            = resultSubject.subscribe(onUpdate, {}, {})

    private fun <T> applySchedulers(s: Single<T>) = s.subscribeOn(scheduler)

    private fun <T> applySchedulers(s: Observable<T>) = s.subscribeOn(scheduler)

    private fun <T> Single<T>.applyBehaviors(behaviors: Array<out SingleBehavior>, params: BehaviorParams) =
            let { behaviors.foldRight(it, { transformer, acc -> transformer.transform(acc, , params) }) }

    private fun <T> Observable<T>.applyBehaviors(behaviors: Array<out ObservableBehavior>, params: BehaviorParams) =
            let { behaviors.foldRight(it, { transformer, acc -> transformer.transform(acc, , params) }) }

    private open class DefaultLocationListener(protected val emitter: Emitter<LocationEvent>) : LocationListener {
        override fun onLocationChanged(location: Location) {
            emitter.onNext(LocationEvent.LocationChangedEvent(location))
        }

        override fun onProviderDisabled(provider: String) {
            emitter.onNext(LocationEvent.ProviderEnableStatusEvent(provider, false))
        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle?) {
            emitter.onNext(LocationEvent.StatusChangedEvent(provider, status, extras))
        }

        override fun onProviderEnabled(provider: String) {
            emitter.onNext(LocationEvent.ProviderEnableStatusEvent(provider, true))
        }
    }

    private class SingleLocationListener(emitter: Emitter<LocationEvent>) : DefaultLocationListener(emitter) {
        override fun onLocationChanged(location: Location) {
            super.onLocationChanged(location)
            emitter.onCompleted()
        }
    }
}


